{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "19f49e56",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Goals of notebook:\n",
    "## 1. Access ChEMBL database\n",
    "## 2. Shortlist active compounds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "114ffc4c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\kheef\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\chembl_webresource_client\\__init__.py:4: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.\n",
      "  __version__ = __import__('pkg_resources').get_distribution('chembl_webresource_client').version\n"
     ]
    }
   ],
   "source": [
    "# !pip install chembl_webresource_client\n",
    "from chembl_webresource_client.new_client import new_client\n",
    "import pandas as pd\n",
    "import re\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2d32de08",
   "metadata": {},
   "outputs": [],
   "source": [
    "#connecting to chemBL\n",
    "activity = new_client.activity\n",
    "assay = new_client.assay\n",
    "molecule = new_client.molecule\n",
    "target = new_client.target"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "70ce3529",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 1 KRAS targets\n"
     ]
    }
   ],
   "source": [
    "kras_targets = target.filter(\n",
    "    target_synonym__icontains=\"KRAS\",\n",
    "    target_type=\"SINGLE PROTEIN\",\n",
    "    organism=\"Homo sapiens\"\n",
    ")\n",
    "\n",
    "kras_target_ids = [\n",
    "    t[\"target_chembl_id\"]\n",
    "    for t in kras_targets\n",
    "]\n",
    "\n",
    "print(f\"Found {len(kras_target_ids)} KRAS targets\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "295d5590",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target_chembl_id</th>\n",
       "      <th>pref_name</th>\n",
       "      <th>organism</th>\n",
       "      <th>target_type</th>\n",
       "      <th>uniprot_accessions</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHEMBL2189121</td>\n",
       "      <td>GTPase KRas</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>P01116</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  target_chembl_id    pref_name      organism     target_type  \\\n",
       "0    CHEMBL2189121  GTPase KRas  Homo sapiens  SINGLE PROTEIN   \n",
       "\n",
       "  uniprot_accessions  \n",
       "0             P01116  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Convert to DataFrame for inspection\n",
    "df_targets = pd.DataFrame([\n",
    "    {\n",
    "        \"target_chembl_id\": t[\"target_chembl_id\"],\n",
    "        \"pref_name\": t[\"pref_name\"],\n",
    "        \"organism\": t[\"organism\"],\n",
    "        \"target_type\": t[\"target_type\"],\n",
    "        \"uniprot_accessions\": t.get(\"target_components\", [{}])[0].get(\"accession\")\n",
    "    }\n",
    "    for t in kras_targets\n",
    "])\n",
    "\n",
    "df_targets.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1abb15bb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Candidate G12D assays: 85\n"
     ]
    }
   ],
   "source": [
    "#filter for those targeting G12D mutation and confidence score >= 8\n",
    "\n",
    "candidate_assays = []\n",
    "\n",
    "for tid in kras_target_ids:\n",
    "    assays = assay.filter(target_chembl_id=tid)\n",
    "    for a in assays:\n",
    "        desc = (a.get(\"description\") or \"\").lower()\n",
    "        if \"g12d\" in desc:\n",
    "            candidate_assays.append({\n",
    "                \"assay_chembl_id\": a[\"assay_chembl_id\"],\n",
    "                \"assay_type\": a.get(\"assay_type\"),\n",
    "                \"description\": a.get(\"description\")\n",
    "            })\n",
    "\n",
    "print(f\"Candidate G12D assays: {len(candidate_assays)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a762928d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assay_chembl_id</th>\n",
       "      <th>assay_type</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>CHEMBL5161453</td>\n",
       "      <td>B</td>\n",
       "      <td>Inhibition of KRAS G12D mutant downstream sign...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>CHEMBL5553095</td>\n",
       "      <td>B</td>\n",
       "      <td>Binding affinity to biotinylated KRAS G12D mut...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>CHEMBL5217977</td>\n",
       "      <td>B</td>\n",
       "      <td>Binding affinity to KRAS G12D mutant (unknown ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHEMBL2399317</td>\n",
       "      <td>B</td>\n",
       "      <td>Binding affinity to KRAS G12D mutant (unknown ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>CHEMBL5737727</td>\n",
       "      <td>B</td>\n",
       "      <td>Ras GTP Binding Domain Inhibition Assay: The f...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   assay_chembl_id assay_type  \\\n",
       "49   CHEMBL5161453          B   \n",
       "75   CHEMBL5553095          B   \n",
       "50   CHEMBL5217977          B   \n",
       "1    CHEMBL2399317          B   \n",
       "78   CHEMBL5737727          B   \n",
       "\n",
       "                                          description  \n",
       "49  Inhibition of KRAS G12D mutant downstream sign...  \n",
       "75  Binding affinity to biotinylated KRAS G12D mut...  \n",
       "50  Binding affinity to KRAS G12D mutant (unknown ...  \n",
       "1   Binding affinity to KRAS G12D mutant (unknown ...  \n",
       "78  Ras GTP Binding Domain Inhibition Assay: The f...  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Convert to DataFrame for inspection\n",
    "\n",
    "df_assays = pd.DataFrame([\n",
    "    {\n",
    "        \"assay_chembl_id\": c[\"assay_chembl_id\"],\n",
    "        \"assay_type\": c[\"assay_type\"],\n",
    "        \"description\": c[\"description\"]\n",
    "    }\n",
    "    for c in candidate_assays\n",
    "])\n",
    "\n",
    "df_assays.sample(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "79fcfd6f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['B']\n"
     ]
    }
   ],
   "source": [
    "#all assay types seem labelled as binding\n",
    "\n",
    "assay_types = df_assays[\"assay_type\"].unique()\n",
    "print(assay_types)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3f47d9c9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(83,)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "descriptions = df_assays[\"description\"].unique()\n",
    "descriptions.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d138ce4c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Binding affinity to K-Ras G12D mutant-GDP complex\n",
      "Binding affinity to KRAS G12D mutant (unknown origin)\n",
      "Inhibition of biotinylated K-Ras G12D mutant (unknown origin) assessed as reduction in SOS1 (564 to 1049 residues)-mediated BODIPY-GDP-GTP exchange after 1 hr by TR-FRET assay\n",
      "Inhibition of biotinylated K-Ras G12D mutant (unknown origin) assessed as reduction in SOS1 (564 to 1049 residues)-mediated BODIPY-GDP-GTP exchange after 1 hr in presence of DTT by TR-FRET assay\n",
      "Binding affinity to human N-terminal His/AVi-tagged biotinylated GTPase KRas G12D mutant (1 to 169 residues) expressed in Escherichia coli BL21 (DE3) in presence of GDP by SPR assay\n",
      "Binding affinity to human N-terminal His/AVi-tagged biotinylated GTPase KRas G12D mutant (1 to 169 residues) expressed in Escherichia coli BL21 (DE3) in presence of GTP by SPR assay\n",
      "Binding affinity to human N-terminal His/AVi-tagged biotinylated GTPase KRas G12D mutant (1 to 169 residues) expressed in Escherichia coli BL21 (DE3) assessed as association rate constant in presence of GDP by SPR assay\n",
      "Binding affinity to human N-terminal His/AVi-tagged biotinylated GTPase KRas G12D mutant (1 to 169 residues) expressed in Escherichia coli BL21 (DE3) assessed as association rate constant in presence of GTP by SRP assay\n",
      "Binding affinity to human N-terminal His/AVi-tagged biotinylated GTPase KRas G12D mutant (1 to 169 residues) expressed in Escherichia coli BL21 (DE3) assessed as dissociation rate constant in presence of GDP by SPR assay\n",
      "Binding affinity to human N-terminal His/AVi-tagged biotinylated GTPase KRas G12D mutant (1 to 169 residues) expressed in Escherichia coli BL21 (DE3) assessed as dissociation rate constant in presence of GTP by SRP assay\n",
      "Binding affinity to human N-terminal His/AVi-tagged biotinylated GTPase KRas G12D mutant (1 to 169 residues) expressed in Escherichia coli BL21 (DE3) assessed as dissociation half life in presence of GDP by SPR assay\n",
      "Binding affinity to human N-terminal His/AVi-tagged biotinylated GTPase KRas G12D mutant (1 to 169 residues) expressed in Escherichia coli BL21 (DE3) assessed as dissociation half life in presence of GTP by SRP assay\n",
      "Inhibition of KRAS G12D mutant in human AGS cells assessed as reduction in ERK phosphorylation measured after 3 hrs by Western blot analysis\n",
      "Binding affinity to 15N-labelled GDP bound KRas G12D mutant ( 1 to 169 residues) (unknown origin) expressed in Escherichia coli Rosetta 2 (DE3) cells by 1H/15N- HSQC spectroscopy\n",
      "Binding affinity to GCP-KRAS G12D (unknown origin) by isothermal titration calorimetry\n",
      "Inhibition of C-terminal Avi-tagged biotinylated KRas G12D mutant (1 to 169 residues) (unknown origin) expressed in Escherichia coli assessed as reduction in KRas G12D mutant-SOS1 (564 to 1049 residues) protein-protein interaction incubated for 60 mins by TR-FRET LANCE assay\n",
      "Binding affinity to GTP-KRAS G12D (unknown origin) by isothermal titration calorimetry\n",
      "Inhibition of N-terminal 6His-tagged/C-terminal Avi-tagged biotinylated GDP-bound human KRas G12D mutant (1 to 169 residues) assessed as reduction in KRas G12D mutant-SOS1 (564 to 1049 residues) protein-protein interaction by alpha screen assay\n",
      "Binding affinity to human KRAS G12D mutant (1 to 169 residues) expressed in Escherichia coli BL21(DE3) by HSQC NMR spectroscopy\n",
      "Binding affinity to human GCP-KRAS G12D mutant (1 to 169 residues) expressed in Escherichia coli BL21(DE3) by isothermal titration calorimetry assay\n",
      "Binding affinity to KRAS G12D mutant (unknown origin) using FAM-labelled peptide as substrate in absence of Tween20 by fluorescence polarization assay\n",
      "Binding affinity to KRAS G12D mutant (unknown origin) upto 1 uM by surface plasmon resonance analysis\n",
      "Binding affinity to KRAS G12D mutant (unknown origin) by isothermal titration calorimetry\n",
      "Binding affinity to KRAS G12D mutant (unknown origin) assessed as dissociation constant by isothermal titration calorimetry\n",
      "Binding affinity to KRAS G12D mutant (unknown origin) assessed as change in melting temperature by SYPRO orange dye based thermal shift assay\n",
      "Binding affinity to KRAS G12D mutant (unknown origin) assessed as absence in deuterium uptake by hydrogen/deuterium exchange/mass spectrometry\n",
      "Binding affinity to KRAS G12D mutant (unknown origin) using FAM-labelled peptide as substrate in presence of Tween20 by fluorescence polarization assay\n",
      "Binding affinity to KRAS G12D mutant (unknown origin) in presence of FAM-labelled peptide at 1 uM by surface plasmon resonance analysis\n",
      "Binding affinity to GDP-bound KRAS G12D mutant (unknown origin) assessed as bound fraction at 100 uM incubated for 20 hrs by RF-MS analysis relative to control\n",
      "Inhibition of recombinant his-tagged KRAS G12D mutant (unknown origin) expressed in Escherichia coli BL21(DE3) assessed as inhibition of GDP-GTP exchange rate in presence of recombinant SOS1/GMPPNP incubated for 1 hr by BODIPY fluorescence assay\n",
      "Displacement of BODIPY-FL-GD from human KRAS G12D (Met1 to Lys169 residues) incubated for over 1 hr by TR-FRET assay\n",
      "Binding affinity to recombinant human KRAS G12D mutant incubated for 0.5 hrs by ELISA\n",
      "Inhibition of KRas G12D mutant (unknown origin) expressed in HEK293T cells coexpressing GFP-fused CRAF RBD assessed as reduction in Ras-Raf interaction using coelenterazine 400a as substrate incubated for 20 hrs by BRET2 assay\n",
      "Inhibition of KRas G12D mutant (unknown origin) expressed in HEK293T cells coexpressing GFP-fused CRAF RBD assessed as reduction in Ras-Raf interaction using coelenterazine 400a as substrate up to 15 uM incubated for 20 hrs by BRET2 assay\n",
      "Displacement of Cy5-labelled tracer from biotinylated GDP-loaded human recombinant KRAS G12D mutant (1 to 169 residues) measured after 60 mins by TR-FRET assay\n",
      "Inhibition of KRAS G12D mutant in human AGS cells assessed as reduction in ERK phosphorylation measured after 3 hrs by In-Cell Western assay\n",
      "Inhibition of human GDP-bound KRAS G12D mutant assessed as inhibition of SOS1-catalyzed nucleotide exchange by measuring KRAS G12D mutant/GTP/c-Raf1 complex formation incubated for 120 mins followed by incubation with SOS1 and GTP for 60 mins followed by addition of c-Raf1 for 60 mins and measured after 60 mins by AlphaLISA assay\n",
      "Inhibition of human His-Avi-TEV-tagged KRAS G12D mutant expressed in Escherichia coli BL21(DE3) by TR-FRET assay\n",
      "Binding affinity to GppNHp-bound Avidin-tagged KRAS G12D mutant (2 to 169 residues) (unknown origin) up to 100 uM by surface plasmon resonance analysis\n",
      "Binding affinity to KRAS G12D mutant (unknown origin) measured after 30 mins by TAMRA-N3 based In-gel fluorescence scanning analysis\n",
      "Inhibition of recombinant biotinylated KRAS G12D mutant (unknown origin) assessed as inhibition of SOS-catalyzed nucleotide exchange preincubated for 60 mins followed by addition of human SOS protein and unlabeled GTP and measured after 60 mins by TR-FRET analysis\n",
      "Inhibition of KRAS in human ASPC1 cells harboring KRAS G12D mutant assessed as inhibition of ERK phosphorylation incubated for 2 hrs by AlphaScreen-based LDH membrane integrity assay\n",
      "Inhibition of KRAS in human ASPC1 cells harboring KRAS G12D mutant assessed as inhibition of ERK phosphorylation incubated for 18 hrs by AlphaScreen-based LDH membrane integrity assay\n",
      "Binding affinity to GDP-bound KRAS G12D mutant (unknown origin) assessed as dissociation constant\n",
      "Inhibition of KRAS G12D mutant (unknown origin) downstream signalling in human SW1990 cells assessed as suppression of Raf phosphorylation incubated for 24 hrs by pull down based Western blotting analysis\n",
      "Binding affinity to human his-tagged KRAS G12D mutant assessed as dissociation constant by isothermal titration calorimetry\n",
      "Binding affinity to human his-tagged KRAS G12D mutant assessed as change in enthalpy by isothermal titration calorimetry\n",
      "Binding affinity to human his-tagged KRAS G12D mutant assessed as gibbs free energy change by isothermal titration calorimetry\n",
      "Binding affinity to human his-tagged KRAS G12D mutant assessed as change in entropy by isothermal titration calorimetry\n",
      "Inhibition of KRAS G12D mutant downstream signalling in human SW1990 cells assessed as suppression of ERK phosphorylation incubated for 24 hrs by pull down based Western blotting analysis\n",
      "Binding affinity to KRAS G12D mutant (unknown origin) assessed as dissociation constant by HSQC NMR spectroscopy analysis\n",
      "Binding affinity to human GTPase KRas G12D mutant assessed as dissociation constant by ELISA\n",
      "Inhibition of KRAS G12D mutant in human ASPC1 cells assessed as reduction in ERK phosphorylation after 2 hrs by Western blot analysis\n",
      "Inhibition of KRAS G12D mutant in human ASPC1 cells assessed as reduction in AKT phosphorylation after 2 hrs by Western blot analysis\n",
      "Binding affinity to 6His-tagged KRAS G12D mutant (unknown origin) assessed as change in melting temperature at 10 uM by TSA (Rvb = 0 degreeC)\n",
      "Binding affinity to GDP-bound KRAS G12D mutant (unknown origin) by 1H/15N HSQC spectra analysis\n",
      "Inhibition of recombinant GDP-bound KRAS G12D mutant (unknown origin) expressed in Escherichia coli Rosetta 2 (DE3) assessed as inhibition of SOS mediated nucleotide exchange at 1 mM relative to control\n",
      "Inhibition of human KRAS G12D mutant in human AGS cells assessed as reduction in phosphorylated ERK level incubated for 3 hrs\n",
      "Inhibition of KRAS G12D mutant (unknown origin) preincubated for 4 hrs followed by SOS1 addition and measured after 2 hrs by fluorescence based analysis\n",
      "Inhibition of GTP bound KRAS G12D mutant in human GP2d cells assessed as reduction in pERK level incubated for 3 hrs\n",
      "Inhibition of His-tagged KRAS G12D mutant (unknown origin) using biotinylated KRPep-2d as substrate preincubated for 30 mins followed by substrate addition and measured after 2 hrs by LANCE TR-FRET assay\n",
      "Inhibition of N-terminal His-tagged GDP-bound KRAS G12D mutant (1 to 169 residues) (unknown origin) assessed as inhibition of SOS mediated nucleotide exchange preincubated for 2 hrs followed by addition of SOS protein and unlabeled GTP measured after 30 mins by Alphascreen assay\n",
      "Inhibition of KRAS G12D mutant in human ASPC1 cells assessed as reduction in ERK1/2 phosphorylation level incubated for 2 hrs by MSD assay\n",
      "Protac activity at VHL/KRAS G12D in human ASPC1 cells assessed as induction of KRAS G12D protein degradation at 1 uM incubated for 24 hrs by luminescence based analysis relative to control\n",
      "Inhibition of KRAS G12D mutant in human PANC-1 cells assessed as reduction in phosphorylation of ERK 1/2 level incubated for 3 hrs\n",
      "Inhibition of KRAS G12D mutant (unknown origin) assessed as inhibition of SOS1 mediated nucleotide exchange of bodipy-GDP for GppNHp incubated for 4 hrs by fluorescence based analysis\n",
      "Inhibition of KRAS G12D mutant in human HPAF-II cells assessed as reduction in ERK phosphorylation by Western blot analysis\n",
      "Inhibition of GDP-bound biotinylated-KRAS G12D mutant (1 to 169 residues) (unknown origin) assessed as suppression of SOS1-mediated nucleotide exchange by measuring reduction in KRAS G12D mutant-RAF RBD complex formation preincubated for 60 mins followed by SOS1 addition and further incubated for 60 mins and subsequent addition of KRAS G12D mutant-RAF RBD mixture measured after 30 to 60 mins by FRET assay\n",
      "Thermal stabilization of GDP-bound KRAS G12D mutant (1 to 169 residues) (unknown origin) assessed as melting temperature incubated for 1 hrs by differential scanning fluorimetry assay\n",
      "Inhibition of KRAS G12D mutant in human ASPC1 cells assessed as reduction in ERK phosphorylation measured after 4 hrs by HTRF assay\n",
      "Inhibition of KRAS G12D mutant (unknown origin) preincubated for 15 mins followed by SOS1/EDA-GTP-DY-647P1 addition measured after 30 mins by HTRF assay\n",
      "Inhibition of KRAS G12D mutant in human PANC-1 cells incubated for 24 hrs by Western blot analysis\n",
      "Inhibition of KRAS G12D mutant in human PANC-1 cells at 2.5 uM by Western blot analysis\n",
      "Binding affinity to biotinylated KRAS G12D mutant (unknown origin) incubated for 120 mins by SPR analysis\n",
      "Binding affinity to biotinylated KRAS G12D mutant (unknown origin) assessed as Kon incubated for 120 mins by SPR analysis\n",
      "Binding affinity to biotinylated KRAS G12D mutant (unknown origin) assessed as Koff incubated for 120 mins by SPR analysis\n",
      "Binding Domain Inhibition Assay: Buffer-I:    50 mM Tris, pH 7.5    150 mM NaCl (optional)    1 mM MgCl2    1 mM DTT.    KRas G12D mutant protein was expressed as a His-tagged protein. Purified His-KRas G12D protein was diluted in buffer-I to a final concentration of 3-10 μg/ml.    200 μl of the diluted His-KRas G12D protein was added to a nickel coated 96 well plate and incubated overnight at 4° C.    The next day, wells were washed 3× in 200 μl of Buffer-I.    Then 200 μl of Buffer-I were added to each well in the presence of 1% DMSO.    Tested compounds were added to the protein-coated wells at a concentration of 20 μM, and incubated for 3 hours at room temperature. While performing IC50 measurements a serial dilution of all tested concentrations was prepared.Then 22 μl of Cy3-GTP or Cy5-GTP was added to each well. The labeled GTP was incubated for 45 min. at room temperature.\n",
      "Inhibition Assay: Buffer-I:    50 mM Tris, pH 7.5    150 mM NaCl (optional)    1 mM MgCl2    1 mM DTT.    KRas G12D mutant protein was expressed as a His-tagged protein. Purified His-KRas G12D protein was diluted in buffer-I to a final concentration of 3-10 μg/ml.    200 μl of the diluted His-KRas G12D protein was added to a nickel coated 96 well plate and incubated overnight at 4° C.    The next day, wells were washed 3× in 200 μl of Buffer-I.    Then 200 μl of Buffer-I were added to each well in the presence of 1% DMSO.    Tested compounds were added to the protein-coated wells at a concentration of 20 μM, and incubated for 3 hours at room temperature. While performing IC50 measurements a serial dilution of all tested concentrations was prepared.Then 22 μl of Cy3-GTP or Cy5-GTP was added to each well. The labeled GTP was incubated for 45 min. at room temperature.    Following GTP incubation, wells were washed 3× in Buffer-I, and 200 μl of Buffer-I were added to each well.    Following washes, the amount of bound labeled-GTP was measured with an Eppendorf AF2200 plate reader.\n",
      "Ras GTP Binding Domain Inhibition Assay: The following method was developed as specific assay for KRas G12D mutant protein.Buffer-I:    50 mM Tris, pH 7.5    150 mM NaCl (optional)    1 mM MgCl2    1 mM DTT.    KRas G12D mutant protein was expressed as a His-tagged protein. Purified His-KRas G12D protein was diluted in buffer-I to a final concentration of 3-10 μg/ml.    200 μl of the diluted His-KRas G12D protein was added to a nickel coated 96 well plate and incubated overnight at 4° C.    The next day, wells were washed 3× in 200 μl of Buffer-I.    Then 200 μl of Buffer-I were added to each well in the presence of 1% DMSO.    Tested compounds were added to the protein-coated wells at a concentration of 20 μM, and incubated for 3 hours at room temperature. While performing IC50 measurements a serial dilution of all tested concentrations was prepared.Then 22 μl of Cy3-GTP or Cy5-GTP was added to each well. The labeled GTP was incubated for 45 min. at room temperature.    Following GTP incubation, wells were washed 3× in Buffer-I, and 200 μl of Buffer-I were added to each well.    Following washes, the amount of bound labeled-GTP was measured with an Eppendorf AF2200 plate reader.By substituting KRas G12D mutant protein with KRas G12C mutant, KRas wild type, KRas Q61H mutant, KRas G12D/Q61H double mutant or KRas G12C/Q61H double mutant under the assay conditions described above, KRas G12C mutant KRas wild type, KRas Q61H mutant, KRas G12D/Q61H double mutant and KRas G12C/Q61H double mutant inhibition, respectively, were each determined.\n",
      "KRas G12D Surface Plasmon Resonance (SPR) Binding Assay: Table 1: Briefly, 1 L of 1.05× HBS-Mg buffer (262.5 mM Bioultra Hepes, pH 7.5, 157.5 mM NaCl, 105 mM MgCl2, 0.525 mM TCEP, 0.0305% Brij-35) was prepared and filter sterilized using a 0.22 μm bottle top filter. Approximately 50 mL of 1.05× HBS-Mg buffer was removed and saved for future dilutions. A 50 mL aliquot of DMSO (Sigma Aldrich DMSO Lot. #SHBK2079) was added and continued to stir for 10 minutes, creating the final 1.0× HBS-Mg buffer (250 mM Bioultra Hepes pH 7.5, 150 mM NaCl, 100 mM MgCl2, 0.5mM TCEP, 0.03% Brij-35).\n",
      "KRas G12D Binding Assay: Table 2: The ability of a compound to bind to KRAS G12D was measured using a TR-FRET displacement assay. Biotinylated GDP-loaded recombinant human KRAS G12D (corresponding to amino acids 1:169, produced at Array BioPharma) was incubated with a custom-made Cy5 labelled tracer, europium labelled streptavidin and compound (2% DMSO final) in buffer (50 mM HEPES [pH 7.5], 5mM MgCl2, 0.005% Tween-20 & 1 mM DTT). After a 60 minute incubation at 22° C., the reaction was measured using a PerkinElmer EnVision multimode plate reader via TR-FRET dual wavelength detection, and the percent of control (POC) calculated using a ratiometric emission factor. 100 POC is determined using no test compound and 0 POC is determined using a concentration of control compound that completely inhibits binding of the tracer to KRAS. The POC values were fit to a 4-parameter logistic curve and the IC50 value was determined as the concentration where the curve crosses 50 POC.\n",
      "Ras GTP Binding Domain Inhibition Assay: The following method was developed as specific assay for KRas G12D mutant protein.Buffer-I:50 mM Tris, pH 7.5150 mM NaCl (optional)1 mM MgCl21 mM DTT.KRas G12D mutant protein was expressed as a His-tagged protein. Purified His-KRas G12D protein was diluted in buffer-I to a final concentration of 3-10 μg/ml.200 μl of the diluted His-KRas G12D protein was added to a nickel coated 96 well plate and incubated overnight at 4° C.The next day, wells were washed 3× in 200 μl of Buffer-I.Then 200 μl of Buffer-I were added to each well in the presence of 1% DMSO.Tested compounds were added to the protein-coated wells at a concentration of 20 μM, and incubated for 3 hours at room temperature. While performing IC50 measurements a serial dilution of all tested concentrations was prepared.Then 22 μl of Cy3-GTP or Cy5-GTP was added to each well. The labeled GTP was incubated for 45 min. at room temperature.Following GTP incubation, wells were washed 3× in Buffer-I, and 200 μl of Buffer-I were added to each well.Following washes, the amount of bound labeled-GTP was measured with an Eppendorf AF2200 plate reader.\n",
      "Biochemical Assay: A variety of Ras proteins may be inhibited by compounds of the present invention (e.g., K-Ras, N-Ras, H-Ras, and mutants thereof at positions 12, 13 and 61, such as G12C, G12D, G12V, G12S, G13C, G13D, and Q61L, and others described herein). The purpose of this biochemical assay is to measure the ability of test compounds to covalently label nucleotide-loaded K-Ras isoforms. In assay buffer containing 12.5 mM HEPES pH 7.4, 75 mM NaCl, 1 mM MgCl2, 1 mM BME, 5 uM Cyclophilin A and 2 uM test compound, a 5 uM stock of GMP-PNP-loaded K-Ras (1-169) G12C is diluted 10-fold to yield a final concentration of 0.5 uM; with final sample volume being 100 uL. The sample is incubated at 25�� C. for a time period of up to 24 hours prior to quenching by the addition of 10 uL of 5% Formic Acid. Quenched samples are centrifuged at 15000 rpm for 15 minutes in a benchtop centrifuge before injecting a 10 uL aliquot onto a reverse phase C4 column and eluting into the mass spectrometer with an increasing acetonitrile gradient in the mobile phase. Analysis of raw data may be carried out using Waters MassLynx MS software, with % bound calculated from the deconvoluted protein peaks for labeled and unlabeled K-Ras.\n"
     ]
    }
   ],
   "source": [
    "#can we filter out cell based assays, non-allosteric site bindings and covalent inhibitors based on description?\n",
    "\n",
    "for d in descriptions:\n",
    "  print(d)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7ffffdb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Assays to leave OUT\n",
    "## Cellular assays"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9849280",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(65, 3)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 1. Define your exclusion keywords\n",
    "exclusion_keywords = [\n",
    "    'cell'\n",
    "]\n",
    "\n",
    "# 2. Create the filtering pattern\n",
    "pattern = '|'.join(exclusion_keywords)\n",
    "\n",
    "# 3. Apply the filter to your dataframe\n",
    "## Assumes your dataframe is 'df' and the column is 'assay_description'\n",
    "descriptions_filtered = df_assays[~df_assays['description'].str.contains(pattern, case=False, na=False)].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "953359e4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Inhibition of KRAS G12D mutant in human AGS cells assessed as reduction in ERK phosphorylation measured after 3 hrs by Western blot analysis\n",
      "Binding affinity to 15N-labelled GDP bound KRas G12D mutant ( 1 to 169 residues) (unknown origin) expressed in Escherichia coli Rosetta 2 (DE3) cells by 1H/15N- HSQC spectroscopy\n",
      "Inhibition of KRas G12D mutant (unknown origin) expressed in HEK293T cells coexpressing GFP-fused CRAF RBD assessed as reduction in Ras-Raf interaction using coelenterazine 400a as substrate incubated for 20 hrs by BRET2 assay\n",
      "Inhibition of KRas G12D mutant (unknown origin) expressed in HEK293T cells coexpressing GFP-fused CRAF RBD assessed as reduction in Ras-Raf interaction using coelenterazine 400a as substrate up to 15 uM incubated for 20 hrs by BRET2 assay\n",
      "Inhibition of KRAS G12D mutant in human AGS cells assessed as reduction in ERK phosphorylation measured after 3 hrs by In-Cell Western assay\n",
      "Inhibition of KRAS in human ASPC1 cells harboring KRAS G12D mutant assessed as inhibition of ERK phosphorylation incubated for 2 hrs by AlphaScreen-based LDH membrane integrity assay\n",
      "Inhibition of KRAS in human ASPC1 cells harboring KRAS G12D mutant assessed as inhibition of ERK phosphorylation incubated for 18 hrs by AlphaScreen-based LDH membrane integrity assay\n",
      "Inhibition of KRAS G12D mutant (unknown origin) downstream signalling in human SW1990 cells assessed as suppression of Raf phosphorylation incubated for 24 hrs by pull down based Western blotting analysis\n",
      "Inhibition of KRAS G12D mutant downstream signalling in human SW1990 cells assessed as suppression of ERK phosphorylation incubated for 24 hrs by pull down based Western blotting analysis\n",
      "Inhibition of KRAS G12D mutant in human ASPC1 cells assessed as reduction in ERK phosphorylation after 2 hrs by Western blot analysis\n",
      "Inhibition of KRAS G12D mutant in human ASPC1 cells assessed as reduction in AKT phosphorylation after 2 hrs by Western blot analysis\n",
      "Inhibition of human KRAS G12D mutant in human AGS cells assessed as reduction in phosphorylated ERK level incubated for 3 hrs\n",
      "Inhibition of GTP bound KRAS G12D mutant in human GP2d cells assessed as reduction in pERK level incubated for 3 hrs\n",
      "Inhibition of KRAS G12D mutant in human ASPC1 cells assessed as reduction in ERK1/2 phosphorylation level incubated for 2 hrs by MSD assay\n",
      "Protac activity at VHL/KRAS G12D in human ASPC1 cells assessed as induction of KRAS G12D protein degradation at 1 uM incubated for 24 hrs by luminescence based analysis relative to control\n",
      "Inhibition of KRAS G12D mutant in human PANC-1 cells assessed as reduction in phosphorylation of ERK 1/2 level incubated for 3 hrs\n",
      "Inhibition of KRAS G12D mutant in human HPAF-II cells assessed as reduction in ERK phosphorylation by Western blot analysis\n",
      "Inhibition of KRAS G12D mutant in human ASPC1 cells assessed as reduction in ERK phosphorylation measured after 4 hrs by HTRF assay\n",
      "Inhibition of KRAS G12D mutant in human PANC-1 cells incubated for 24 hrs by Western blot analysis\n",
      "Inhibition of KRAS G12D mutant in human PANC-1 cells at 2.5 uM by Western blot analysis\n"
     ]
    }
   ],
   "source": [
    "descriptions_filtered_test = df_assays[df_assays['description'].str.contains(pattern, case=False, na=False)].copy()\n",
    "\n",
    "descriptions_filtered_test.shape\n",
    "for d in descriptions_filtered_test['description']:\n",
    "  print(d)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87069018",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 4. Optional: Create a 'Source' tag for easier analysis later\n",
    "df_filtered['assay_type'] = df_filtered['assay_description'].apply(\n",
    "    lambda x: 'Cellular' if any(cell in x.upper() for cell in ['CELL', 'AGS', 'ASPC1', 'PANC-1', 'GP2D']) else 'Biochemical'\n",
    ")\n",
    "\n",
    "print(f\"Original assays: {len(df)}\")\n",
    "print(f\"Cleaned assays: {len(df_filtered)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c841bbec",
   "metadata": {},
   "outputs": [],
   "source": [
    "activities_query = activity.filter(\n",
    "    assay_chembl_id__in=df_assays[\"assay_chembl_id\"].unique().tolist()\n",
    "  ).only(\n",
    "    'molecule_chembl_id',\n",
    "    'target_chembl_id',\n",
    "    'standard_type',\n",
    "    'standard_value',\n",
    "    'standard_units',\n",
    "    'standard_relation',\n",
    "    'assay_type',\n",
    "    'confidence_score',\n",
    "    'canonical_smiles',\n",
    "    'data_validity_comment'\n",
    "  )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "392ba608",
   "metadata": {},
   "outputs": [],
   "source": [
    "total_count = len(activities_query)\n",
    "print(f'Total records found on server: {total_count}')\n",
    "\n",
    "activity_list = []\n",
    "for act in tqdm(activities_query, total=total_count, desc=\"Fetching ChEMBL Data\"):\n",
    "  activity_list.append(act)\n",
    "\n",
    "df_activities = pd.DataFrame(activity_list)\n",
    "csv_filename = \"KRAS_G12D_chemBL_raw.csv\"\n",
    "df_activities.to_csv(csv_filename, index=False)\n",
    "print(f'\\nDownload complete!')\n",
    "print(f'Total records saved: {len(df_activities)}')\n",
    "if not df_activities.empty:\n",
    "  print(f'Types of measurements found: {df_activities[\"standard_type\"].unique()}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6de7837b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_activities[['standard_type', 'standard_units']].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8981a60d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_activities[['type', 'units']].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e64f2374",
   "metadata": {},
   "outputs": [],
   "source": [
    "#need to further filter the list of actives here\n",
    "\n",
    "accepted_types = [\"Kd\", \"IC50\"]\n",
    "df_filtered = df_activities[df_activities[\"standard_value\"].notna() & (df_activities[\"relation\"] == \"=\") & df_activities[\"standard_type\"].isin(accepted_types)]\n",
    "df_filtered['standard_value'] = pd.to_numeric(df_filtered['standard_value'], errors='coerce')  # non-numeric -> NaN\n",
    "df_filtered = df_filtered[df_filtered['standard_value'].notna()]  # keep only rows where conversion succeeded"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75061c2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#number of known actives in file\n",
    "\n",
    "print(df_filtered[\"molecule_chembl_id\"].nunique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cfb71691",
   "metadata": {},
   "outputs": [],
   "source": [
    "#only IC50 for plots, scale is nM\n",
    "\n",
    "df_IC50_nM = df_filtered[df_filtered[\"standard_type\"] == \"IC50\"]\n",
    "print(df_IC50_nM[\"molecule_chembl_id\"].nunique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a620be18",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ============================================================\n",
    "# ACTIVITY DISTRIBUTION VISUALIZATION\n",
    "# ============================================================\n",
    "\n",
    "# Visualization\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "fig, axes = plt.subplots(1, 2, figsize=(10, 4))\n",
    "\n",
    "# Histogram\n",
    "ax1 = axes[0]\n",
    "ax1.hist(df_IC50_nM[\"standard_value\"], bins=30, edgecolor='black', alpha=0.7, color='steelblue')\n",
    "ax1.axvline(x=1000.0, color='red', linestyle='--', linewidth=2, label='Activity threshold')\n",
    "ax1.axvline(x=df_IC50_nM[\"standard_value\"].mean(), color='green', linestyle='-', linewidth=2, label=f'Mean: {df_IC50_nM[\"standard_value\"].mean():.2f}')\n",
    "ax1.set_xlabel('IC50 (nM)', fontsize=12)\n",
    "ax1.set_ylabel('Count', fontsize=12)\n",
    "ax1.set_title('Activity Distribution', fontsize=13, fontweight='bold')\n",
    "ax1.legend()\n",
    "\n",
    "# Box plot\n",
    "ax2 = axes[1]\n",
    "bp = ax2.boxplot(df_IC50_nM[\"standard_value\"], patch_artist=True)\n",
    "bp['boxes'][0].set_facecolor('lightblue')\n",
    "ax2.set_ylabel('IC50 (nM)', fontsize=12)\n",
    "ax2.set_title('Activity Box Plot', fontsize=13, fontweight='bold')\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig('figures/activity_distribution.png', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8446663",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
